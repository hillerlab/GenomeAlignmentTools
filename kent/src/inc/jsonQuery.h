/* jsonQuery - simple path syntax for retrieving specific descendants of a jsonElement. */

/* Path examples:
 *
 * "" --> the queried object itself (no-op)
 *
 * "name" --> the value of the "name" field/child of queried object
 *
 * "[0]" --> first element of the queried list
 *
 * "[*]" --> all elements of the queried list
 *
 * "name1.name2[4]" --> 5th element of list that is the value for "name2" child of object
 * that is the value for "name1" child of queried object
 *
 * "primary_snapshot_data.support[id.type=subsnp].submitter_handle" --> dbSNP submitter handles
 * for subsnp (not frequency) submissions
 *
 * See lib/tests/input/json* and lib/tests/expected/json* for more examples with both JSON & paths.
 */

#ifndef JSONQUERY_H
#define JSONQUERY_H

#include "jsonParse.h"

struct slRef *jsonQueryElement(struct jsonElement *el, char *name, char *path, struct lm *lm);
/* Return a ref list of jsonElement descendants of el that match path.
 * name is for error reporting. */

struct slRef *jsonQueryElementList(struct slRef *inList, char *name, char *path, struct lm *lm);
/* Return a ref list of jsonElement descendants matching path of all jsonElements in inList.
 * name is for error reporting. */

char *jsonQueryString(struct jsonElement *el, char *name, char *path, struct lm *lm);
/* Alloc & return the string value at the end of path in el. May be NULL. */

long jsonQueryInt(struct jsonElement *el, char *name, char *path, long defaultVal, struct lm *lm);
/* Return the int value at path in el, or defaultVal if not found. */

boolean jsonQueryBoolean(struct jsonElement *el, char *name, char *path, boolean defaultVal,
                         struct lm *lm);
/* Return the boolean value at path in el, or defaultVal if not found. */

struct slName *jsonQueryStrings(struct jsonElement *el, char *name, char *path, struct lm *lm);
/* Alloc & return a list of string values matching path in el. May be NULL. */

struct slInt *jsonQueryInts(struct jsonElement *el, char *name, char *path, struct lm *lm);
/* Alloc & return a list of int values matching path in el. May be NULL. */

struct slName *jsonQueryStringList(struct slRef *inList, char *name, char *path, struct lm *lm);
/* Alloc & return a list of string values matching path in all elements of inList. May be NULL. */

struct slInt *jsonQueryIntList(struct slRef *inList, char *name, char *path, struct lm *lm);
/* Alloc & return a list of int values matching path in all elements of inList. May be NULL. */

#endif /* JSONQUERY_H */
