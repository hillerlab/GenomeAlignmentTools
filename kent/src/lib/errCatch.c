/* Copyright (C) 2011 The Regents of the University of California 
 * See README in this or parent directory for licensing information. */

/* errCatch - help catch errors so that errAborts aren't
 * fatal, and warn's don't necessarily get printed immediately. 
 * Note that error conditions caught this way will tend to
 * leak resources unless there are additional wrappers. 
 *
 * Typical usage is
 * errCatch = errCatchNew();
 * if (errCatchStart(errCatch))
 *     doFlakyStuff();
 * errCatchEnd(errCatch);
 * if (errCatch->gotError)
 *     warn("%s", errCatch->message->string);
 * errCatchFree(&errCatch); 
 * cleanupFlakyStuff();
 */

#include "common.h"
#include "errAbort.h"
#include "dystring.h"
#include "hash.h"
#include <pthread.h>
#include "errCatch.h"



struct errCatch *errCatchNew()
/* Return new error catching structure. */
{
struct errCatch *errCatch;
AllocVar(errCatch);
errCatch->message = dyStringNew(0);
return errCatch;
}

void errCatchFree(struct errCatch **pErrCatch)
/* Free up resources associated with errCatch */
{
struct errCatch *errCatch = *pErrCatch;
if (errCatch != NULL)
    {
    dyStringFree(&errCatch->message);
    freez(pErrCatch);
    }
}

static struct errCatch **getStack()
/* Return a pointer to the errCatch object stack for the current pthread. */
{
static pthread_mutex_t getStackMutex = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_lock( &getStackMutex );
static struct hash *perThreadStacks = NULL;
pthread_t pid = pthread_self(); //  can be a pointer or a number
// A true integer has function would be nicer, but this will do.  
// Don't safef, theoretically that could abort.
char key[64];
snprintf(key, sizeof(key), "%lld",  ptrToLL(pid));
key[ArraySize(key)-1] = '\0';
if (perThreadStacks == NULL)
    perThreadStacks = hashNew(0);
struct hashEl *hel = hashLookup(perThreadStacks, key);
if (hel == NULL)
    hel = hashAdd(perThreadStacks, key, NULL);
pthread_mutex_unlock( &getStackMutex );
return (struct errCatch **)(&hel->val);
}

static void errCatchAbortHandler()
/* semiAbort */
{
struct errCatch **pErrCatchStack = getStack(), *errCatchStack = *pErrCatchStack;
errCatchStack->gotError = TRUE;
longjmp(errCatchStack->jmpBuf, -1);
}

static void errCatchWarnHandler(char *format, va_list args)
/* Write an error to top of errCatchStack. */
{
struct errCatch **pErrCatchStack = getStack(), *errCatchStack = *pErrCatchStack;
errCatchStack->gotWarning = TRUE;
dyStringVaPrintf(errCatchStack->message, format, args);
dyStringAppendC(errCatchStack->message, '\n');
}

boolean errCatchPushHandlers(struct errCatch *errCatch)
/* Push error handlers.  Not usually called directly. */
{
pushAbortHandler(errCatchAbortHandler);
pushWarnHandler(errCatchWarnHandler);
struct errCatch **pErrCatchStack = getStack();
slAddHead(pErrCatchStack, errCatch);
return TRUE;
}

void errCatchEnd(struct errCatch *errCatch)
/* Restore error handlers and pop self off of catching stack. */
{
popWarnHandler();
popAbortHandler();
struct errCatch **pErrCatchStack = getStack(), *errCatchStack = *pErrCatchStack;
if (errCatch != errCatchStack)
   errAbort("Mismatch between errCatch and errCatchStack");
*pErrCatchStack = errCatch->next;
}

void errCatchReWarn(struct errCatch *errCatch)
/* Re-warn any warnings that happened even though no abort happened 
 * to make them visible. */
{
if (errCatch->gotWarning && !errCatch->gotError)
    {
    warn("%s", errCatch->message->string);
    }
}

boolean errCatchFinish(struct errCatch **pErrCatch)
/* Finish up error catching.  Report error if there is a
 * problem and return FALSE.  If no problem return TRUE.
 * This handles errCatchEnd and errCatchFree. */
{
struct errCatch *errCatch = *pErrCatch;
boolean ok = TRUE;
if (errCatch != NULL)
    {
    errCatchEnd(errCatch);
    if (errCatch->gotError)
	{
	ok = FALSE;
	warn("%s", errCatch->message->string);
	}
    else
	errCatchReWarn(errCatch);
    errCatchFree(pErrCatch);
    }
return ok;
}

