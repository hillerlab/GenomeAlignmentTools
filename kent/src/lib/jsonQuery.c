/* jsonQuery - simple path syntax for retrieving specific descendants of a jsonElement. */

/* Copyright (C) 2018 The Regents of the University of California 
 * See README in this or parent directory for licensing information. */

#include "common.h"
#include "dystring.h"
#include "hash.h"
#include "jsonQuery.h"

static const char *findEndBracket(const char *path)
/* Return a pointer to the right bracket matching the first left bracket that we encounter. */
{
if (path == NULL)
    return NULL;
const char *end = NULL;
int leftCount = 0, rightCount = 0;
int i;
for (i = 0;  path[i] != '\0';  i++)
    {
    if (path[i] == '[')
        leftCount++;
    else if (path[i] == ']')
        {
        rightCount++;
        if (rightCount == leftCount)
            {
            end = path+i;
            break;
            }
        else if (rightCount > leftCount)
            errAbort("findEndBracket: encountered right bracket before left bracket in '%s'",
                     path);
        }
    }
return end;
}

static char *jsonPathPopHead(const char *pathIn, char **retPath, struct lm *lm)
/* Return the first component of pathIn and set retPath to the start of the next component in pathIn.
 * If pathIn is empty/NULL, set retPath to NULL and return the empty string. */
{
if (isEmpty(pathIn))
    {
    *retPath = NULL;
    return lm ? "" : cloneString("");
    }
else
    {
    char *pDot = strchr(pathIn, '.');
    char *pBracket = strchr(pathIn, '[');
    if (pDot && pBracket)
        {
        // Both found -- ignore the second one, handle only the first one.
        if (pDot < pBracket)
            pBracket = NULL;
        else
            pDot = NULL;
        }
    if (pDot)
        {
        if (pDot == pathIn)
            errAbort("jsonPathPopHead: path '%s' should not start with '.'", pathIn);
        *retPath = pDot + 1;
        return lm ? lmCloneStringZ(lm, pathIn, pDot - pathIn)
                  : cloneStringZ(pathIn, pDot - pathIn);
        }
    else if (pBracket)
        {
        if (pBracket == pathIn)
            {
            // This path component is a list index; return contents of []
            const char *pEnd = findEndBracket(pathIn);
            if (!pEnd)
                errAbort("jsonPathPopHead: no matching ']' for '[' in '%s'", pathIn);
            if (pBracket[1] != '*' && !isdigit(pBracket[1]) && !strchr(pBracket, '='))
                errAbort("jsonPathPopHead: contents of '[]' must be '*', a number, "
                         "or a path=val condition (not %s)", pathIn);
            *retPath = (char *)(pEnd + 1);
            if (*retPath[0] == '.')
                *retPath = *retPath + 1;
            return lm ? lmCloneStringZ(lm, pathIn + 1, pEnd - pBracket - 1)
                      : cloneStringZ(pathIn + 1, pEnd - pBracket - 1);
            }
        else
            {
            // The next path component is a list index
            *retPath = pBracket;
            return lm ? lmCloneStringZ(lm, pathIn, pBracket - pathIn)
                      : cloneStringZ(pathIn, pBracket - pathIn);
            }
        }
    else
        {
        // Last component in path
        *retPath = (char *)pathIn + strlen(pathIn);
        return lm ? lmCloneString(lm, pathIn) : cloneString(pathIn);
        }
    }
errAbort("jsonPathPopHead: should have returned a result before this point");
return NULL;
}

// Forward declaration for mutual recursion:
static void rQueryElement(struct jsonElement *elIn, char *name, char *path,
                          struct slRef **pResultList, struct lm *lm);
/* Recursively search for descendants of jsonElements in inList matching path; add jsonElements
 * that match to resultList. */

static void rQueryObject(struct jsonElement *el, char *name, char *id, char *path,
                         struct slRef **pResultList, struct lm *lm)
/* Given a JSON object and a child id, recursively search child for path. */
{
struct hash *hash = jsonObjectVal(el, name);
if (hash)
    {
    struct jsonElement *child = hashFindVal(hash, id);
    if (child)
        rQueryElement(child, name, path, pResultList, lm);
    }
}

static void rQueryList(struct jsonElement *el, char *name, char *id, char *path,
                       struct slRef **pResultList, struct lm *lm)
/* Given a JSON list and a child index (* for all children), recursively search child(ren) if found
 * for path. */
{
struct slRef *list = jsonListVal(el, name);
char *equals = strchr(id, '=');
if (equals)
    {
    // Conditional query; filter list items by condPath=val.
    char *value = equals+1;
    long intVal = atol(value);
    double doubleVal = atof(value);
    boolean booleanVal = FALSE, valIsBoolean = FALSE;
    if (sameString(value, "true") || sameString(value, "TRUE") || sameString(value, "1"))
        {
        booleanVal = TRUE;
        valIsBoolean = TRUE;
        }
    else if (sameString(value, "false") || sameString(value, "FALSE") || sameString(value, "0"))
        {
        booleanVal = FALSE;
        valIsBoolean = TRUE;
        }
    char condPath[strlen(id)+1];
    safencpy(condPath, sizeof condPath, id, (equals - id));
    struct slRef *ref;
    for (ref = list;  ref != NULL;  ref = ref->next)
        {
        struct slRef *condPathResults = NULL;
        rQueryElement(ref->val, name, condPath, &condPathResults, lm);
        struct slRef *resRef;
        for (resRef = condPathResults;  resRef != NULL;  resRef = resRef->next)
            {
            struct jsonElement *resEl = resRef->val;
            boolean matches = FALSE;
            switch (resEl->type)
                {
                case jsonString:
                    matches = sameString(jsonStringVal(resEl, condPath), value);
                    break;
                case jsonNumber:
                    matches = (jsonNumberVal(resEl, condPath) == intVal);
                    break;
                case jsonDouble:
                    matches = (jsonDoubleVal(resEl, condPath) == doubleVal);
                    break;
                case jsonBoolean:
                    if (!valIsBoolean)
                        errAbort("jsonQueryElement: bad conditional value '%s' for boolean", value);
                    matches = (jsonBooleanVal(resEl, condPath) == booleanVal);
                    break;
                case jsonNull:
                    matches = (sameString(value, "NULL") || sameString(value, "null"));
                    break;
                default:
                    errAbort("jsonQueryElement: bad jsonElementType %d for conditional query",
                             resEl->type);
                }
            if (matches)
                {
                rQueryElement(ref->val, name, path, pResultList, lm);
                break;
                }
            }
        }
    }
else
    {
    int idIx = -1;
    if (isdigit(id[0]))
        idIx = atoi(id);
    else if (differentString(id, "*"))
        errAbort("jsonQueryElement: invalid index '%s' for list", id);
    struct slRef *ref;
    int ix;
    for (ref = list, ix = 0;  ref != NULL;  ref = ref->next, ix++)
        {
        if (idIx < 0 || ix == idIx)
            {
            rQueryElement(ref->val, name, path, pResultList, lm);
            if (ix == idIx)
                break;
            }
        }
    }
}

static void rQueryElement(struct jsonElement *elIn, char *name, char *path,
                          struct slRef **pResultList, struct lm *lm)
/* Recursively search for descendants of jsonElements in inList matching path; add jsonElements
 * that match to resultList. */
{
char *pathNext = NULL;
char *id = jsonPathPopHead(path, &pathNext, lm);
struct dyString *dy = dyStringCreate("%s", name);
if (isNotEmpty(id))
    {
    switch (elIn->type)
        {
        case jsonObject:
            {
            dyStringPrintf(dy, ".%s", id);
            rQueryObject(elIn, dy->string, id, pathNext, pResultList, lm);
            break;
            }
        case jsonList:
            {
            dyStringPrintf(dy, "[%s]", id);
            rQueryList(elIn, dy->string, id, pathNext, pResultList, lm);
            break;
            }
        case jsonString:
        case jsonBoolean:
        case jsonNumber:
        case jsonDouble:
        case jsonNull:
            {
            errAbort("jsonQueryElement: got element with scalar type (%d), but children specified "
                     "(%s)", elIn->type, id);
            break;
            }
        default:
            {
            errAbort("jsonQueryElement: invalid type: %d", elIn->type);
            break;
            }
        }
    }
else
    {
    struct slRef *ref;
    if (lm)
        lmAllocVar(lm, ref)
    else
        AllocVar(ref);
    ref->val = elIn;
    slAddHead(pResultList, ref);
    }
if (lm == NULL)
    freez(&id);
dyStringFree(&dy);
}

struct slRef *jsonQueryElementList(struct slRef *inList, char *name, char *path, struct lm *lm)
/* Return a ref list of jsonElement descendants matching path of all jsonElements in inList.
 * name is for error reporting. */
{
struct slRef *resultList = NULL;
struct dyString *dy = dyStringNew(0);
boolean isMult = (inList->next != NULL);
struct slRef *ref;
int ix;
for (ref = inList, ix = 0;  ref != NULL;  ref = ref->next, ix++)
    {
    struct jsonElement *elIn = ref->val;
    if (elIn)
        {
        dyStringClear(dy);
        dyStringPrintf(dy, "%s", name);
        if (isMult)
            dyStringPrintf(dy, "[%d]", ix);
        rQueryElement(elIn, dy->string, path, &resultList, lm);
        }
    }
slReverse(&resultList);
dyStringFree(&dy);
return resultList;
}

struct slRef *jsonQueryElement(struct jsonElement *el, char *name, char *path, struct lm *lm)
/* Return a ref list of jsonElement descendants of el that match path.
 * name is for error reporting. */
{
// Make an slRef wrapper for el and call jsonQueryElementList.
struct slRef elRef = { NULL, el };
return jsonQueryElementList(&elRef, name, path, lm);
}

static struct jsonElement *querySingle(struct jsonElement *el, char *name, char *path, struct lm *lm)
/* Return one jsonElement resulting from searching el for path (or NULL); errAbort if multiple. */
{
struct jsonElement *resEl = NULL;
struct slRef ref = { NULL, el };
struct slRef *resultRef = jsonQueryElementList(&ref, name, path, lm);
if (resultRef)
    {
    if (resultRef->next)
        errAbort("jsonQuerySingle: expected single result but got %d results", slCount(resultRef));
    resEl = resultRef->val;
    }
if (lm == NULL)
    freeMem(resultRef);
return resEl;
}

char *jsonQueryString(struct jsonElement *el, char *name, char *path, struct lm *lm)
/* Alloc & return the string value at the end of path in el. May be NULL. */
{
struct jsonElement *resEl = querySingle(el, name, path, lm);
if (resEl == NULL)
    return NULL;
else if (lm)
    return lmCloneString(lm, jsonStringVal(resEl, path));
else
    return cloneString(jsonStringVal(resEl, path));
}

long jsonQueryInt(struct jsonElement *el, char *name, char *path, long defaultVal, struct lm *lm)
/* Return the int value at path in el, or defaultVal if not found. */
{
struct jsonElement *resEl = querySingle(el, name, path, lm);
long result = resEl ? jsonNumberVal(resEl, path) : defaultVal;
if (lm == NULL)
    freeMem(resEl);
return result;
}

boolean jsonQueryBoolean(struct jsonElement *el, char *name, char *path, boolean defaultVal,
                         struct lm *lm)
/* Return the boolean value at path in el, or defaultVal if not found. */
{
struct jsonElement *resEl = querySingle(el, name, path, lm);
boolean result = resEl ? jsonBooleanVal(resEl, path) : defaultVal;
if (lm == NULL)
    freeMem(resEl);
return result;
}

struct slName *jsonQueryStringList(struct slRef *inList, char *name, char *path, struct lm *lm)
/* Alloc & return a list of string values matching path in all elements of inList. May be NULL. */
{
struct slName *results = NULL;
struct slRef *resultRefs = jsonQueryElementList(inList, name, path, lm);
struct slRef *ref;
for (ref = resultRefs;  ref != NULL;  ref = ref->next)
    {
    struct jsonElement *resEl = ref->val;
    char *string = jsonStringVal(resEl, path);
    struct slName *sln = lm ? lmSlName(lm, string) : slNameNew(string);
    slAddHead(&results, sln);
    }
slReverse(&results);
if (lm == NULL && resultRefs)
    slFreeList(resultRefs);
return results;
}

struct slInt *jsonQueryIntList(struct slRef *inList, char *name, char *path, struct lm *lm)
/* Alloc & return a list of int values matching path in all elements of inList. May be NULL. */
{
struct slInt *results = NULL;
struct slRef *resultRefs = jsonQueryElementList(inList, name, path, lm);
struct slRef *ref;
for (ref = resultRefs;  ref != NULL;  ref = ref->next)
    {
    struct jsonElement *resEl = ref->val;
    int val = jsonNumberVal(resEl, path);
    struct slInt *sli;
    if (lm)
        lmAllocVar(lm, sli)
    else
        AllocVar(sli);
    sli->val = val;
    slAddHead(&results, sli);
    }
slReverse(&results);
if (lm == NULL && resultRefs)
    slFreeList(resultRefs);
return results;
}

struct slName *jsonQueryStrings(struct jsonElement *el, char *name, char *path, struct lm *lm)
/* Alloc & return a list of string values matching path in el. May be NULL. */
{
struct slRef elRef = { NULL, el };
return jsonQueryStringList(&elRef, name, path, lm);
}

struct slInt *jsonQueryInts(struct jsonElement *el, char *name, char *path, struct lm *lm)
/* Alloc & return a list of int values matching path in el. May be NULL. */
{
struct slRef elRef = { NULL, el };
return jsonQueryIntList(&elRef, name, path, lm);
}
