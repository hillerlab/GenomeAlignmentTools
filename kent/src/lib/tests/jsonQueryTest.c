/* jsonQueryTest - Read in JSON and paths from files and print the results. */

/* Copyright (C) 2018 The Regents of the University of California 
 * See README in this or parent directory for licensing information. */

#include "common.h"
#include "errCatch.h"
#include "hash.h"
#include "linefile.h"
#include "options.h"
#include "jsonQuery.h"

void usage()
/* Explain usage and exit. */
{
errAbort(
  "jsonQueryTest - Read in JSON and paths from files and print the results.\n"
  "usage:\n"
  "   jsonQueryTest jsonIn.txt pathIn.txt out.txt\n"
  "jsonIn.txt contains one valid JSON string per line.\n"
  "pathIn.txt contains one valid search path per line.\n"
  "Resulting values will be printed, tab-separated, in out.txt.\n"
  "\n"
  );
}

static struct optionSpec options[] = {
   {NULL, 0},
};

static struct slName *lineFileToList(struct lineFile *lf)
/* Return a list of lines in lf. */
{
struct slName *list = NULL;
char *line;
int size;
while (lineFileNext(lf, &line, &size))
    slAddHead(&list, slNameNewN(line, size));
slReverse(&list);
return list;
}

void shallowPrintEl(FILE *f, struct jsonElement *el, char *name)
/* Print abbreviated version if el is an object or list, or the value if object is a scalar. */
{
if (name != NULL)
    {
    fprintf(f, "\"%s\": ", name);
    }
switch (el->type)
    {
    case jsonObject:
        {
        struct hash *hash = jsonObjectVal(el, "el");
        struct hashEl *helList = hashElListHash(hash);
	fprintf(f, "{");
        if (helList)
            {
            shallowPrintEl(f, helList->val, helList->name);
            if (helList->next)
                fputs(",...", f);
            hashElFreeList(&helList);
            }
        fputc('}', f);
        }
        break;
    case jsonList:
        {
        struct slRef *vals = jsonListVal(el, name);
	fprintf(f, "[");
        if (vals)
            {
            shallowPrintEl(f, vals->val, NULL);
            if (vals->next)
                fputs(",...", f);
            }
        fputc(']', f);
        }
        break;
    case jsonString:
        {
	char *escaped = jsonStringEscape(el->val.jeString);
	fprintf(f, "\"%s\"",  escaped);
	freez(&escaped);
        }
	break;
    case jsonBoolean:
        {
	char *val = (el->val.jeBoolean ? "true" : "false");
	fprintf(f, "%s", val);
        }
	break;
    case jsonNumber:
	fprintf(f, "%ld", el->val.jeNumber);
	break;
    case jsonDouble:
	fprintf(f, "%g", el->val.jeDouble);
        break;
    case jsonNull:
        fprintf(f, "null");
        break;
    default:
        errAbort("shallowPrintEl: invalid type: %d", el->type);
        break;
    }
}

#define MAX_JSON 50

static void jsonQueryTest(char *jsonFile, char *pathFile, char *outFile)
/* Read in JSON and paths from files and print the results. */
{
struct lineFile *jsonLf = lineFileOpen(jsonFile, TRUE);
struct lineFile *pathLf = lineFileOpen(pathFile, TRUE);
FILE *outF = mustOpen(outFile, "w");
struct slName *paths = lineFileToList(pathLf);
char *line;
int size;
while (lineFileNext(jsonLf, &line, &size))
    {
    if (size <= MAX_JSON)
        fprintf(outF, "# JSON: '%s'\n", line);
    else
        fprintf(outF, "# JSON: '%-*s...'\n", MAX_JSON, line);
    struct jsonElement *top = jsonParse(line);
    struct slRef topRef;
    topRef.next = NULL;
    topRef.val = top;
    struct slName *path;
    for (path = paths;  path != NULL;  path = path->next)
        {
        fprintf(outF, "# path: '%s'\n", path->name);
        struct errCatch *errCatch = errCatchNew();
        if (errCatchStart(errCatch))
            {
            struct slRef *refResults = jsonQueryElementList(&topRef, "test JSON", path->name, NULL);
            if (refResults == NULL)
                fputs("NULL\n", outF);
            else
                {
                struct slRef *ref;
                for (ref = refResults;  ref != NULL;  ref = ref->next)
                    {
                    struct jsonElement *el = ref->val;
                    shallowPrintEl(outF, el, NULL);
                    if (ref->next)
                        fputc('\t', outF);
                    }
                fputc('\n', outF);
                }
            }
        errCatchEnd(errCatch);
        if (errCatch->gotError)
            fprintf(outF, "# failed: %s", errCatch->message->string);
        errCatchFree(&errCatch); 
        }
    fputc('\n', outF);
    }
}

int main(int argc, char *argv[])
/* Process command line. */
{
optionInit(&argc, argv, options);
if (argc != 4)
    usage();
jsonQueryTest(argv[1], argv[2], argv[3]);
return 0;
}
