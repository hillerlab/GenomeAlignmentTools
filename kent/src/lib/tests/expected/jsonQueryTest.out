# JSON: '{}'
# path: ''
{}
# path: 'not.found'
NULL
# path: 'name'
NULL
# path: 'id'
NULL
# path: '[1]'
NULL
# path: '[0].a'
NULL
# path: 'books[2]'
NULL
# path: 'books[*].someGuy'
NULL
# path: 'books[100]'
NULL
# path: 'books[0].isTrilogy'
NULL
# path: 'books[3].isTrilogy'
NULL
# path: 'books[number=1].title'
NULL
# path: 'books[title=LOTR:TT].double'
NULL
# path: 'books[double=2.0]'
NULL
# path: 'books[other=null].number'
NULL
# path: 'books[isTrilogy=true]'
NULL
# path: 'books[isTrilogy=false]'
NULL
# path: 'extra.extra[0]'
NULL

# JSON: '"just a string"'
# path: ''
"just a string"
# path: 'not.found'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (not)
# path: 'name'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (name)
# path: 'id'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (id)
# path: '[1]'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (1)
# path: '[0].a'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (0)
# path: 'books[2]'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[*].someGuy'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[100]'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[0].isTrilogy'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[3].isTrilogy'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[number=1].title'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[title=LOTR:TT].double'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[double=2.0]'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[other=null].number'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[isTrilogy=true]'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'books[isTrilogy=false]'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (books)
# path: 'extra.extra[0]'
# failed: jsonQueryElement: got element with scalar type (5), but children specified (extra)

# JSON: '100'
# path: ''
100
# path: 'not.found'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (not)
# path: 'name'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (name)
# path: 'id'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (id)
# path: '[1]'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (1)
# path: '[0].a'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (0)
# path: 'books[2]'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[*].someGuy'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[100]'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[0].isTrilogy'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[3].isTrilogy'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[number=1].title'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[title=LOTR:TT].double'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[double=2.0]'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[other=null].number'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[isTrilogy=true]'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'books[isTrilogy=false]'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (books)
# path: 'extra.extra[0]'
# failed: jsonQueryElement: got element with scalar type (2), but children specified (extra)

# JSON: '[{"a":"two"}, {"a":"things"}]'
# path: ''
[{"a": "two"},...]
# path: 'not.found'
# failed: jsonQueryElement: invalid index 'not' for list
# path: 'name'
# failed: jsonQueryElement: invalid index 'name' for list
# path: 'id'
# failed: jsonQueryElement: invalid index 'id' for list
# path: '[1]'
{"a": "things"}
# path: '[0].a'
"two"
# path: 'books[2]'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[*].someGuy'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[100]'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[0].isTrilogy'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[3].isTrilogy'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[number=1].title'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[title=LOTR:TT].double'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[double=2.0]'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[other=null].number'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[isTrilogy=true]'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'books[isTrilogy=false]'
# failed: jsonQueryElement: invalid index 'books' for list
# path: 'extra.extra[0]'
# failed: jsonQueryElement: invalid index 'extra' for list

# JSON: '{ "id" : 1234, "name" : "z", "books" : [ {"title": "hobbit", "someGuy":"Bilbo", "isTrilogy":false, "number": 0, "double": 0.0, "other": null}, {"title": "LOTR:FOTR", "someGuy":"Gandalf", "isTrilogy":true, "number": 1, "double": 1.0 }, {"title":"LOTR:TT", "someGuy":"Sam", "isTrilogy":true, "number": 2, "double": 2.0 }, {"title":"LOTR:ROTK", "someGuy": "Aragorn", "isTrilogy":true, "number": 3, "double": 3.0, "other": "all done"}], "extra":{ "extra": [3209423, 2354341]} }...'
# path: ''
{"id": 1234,...}
# path: 'not.found'
NULL
# path: 'name'
"z"
# path: 'id'
1234
# path: '[1]'
NULL
# path: '[0].a'
NULL
# path: 'books[2]'
{"title": "LOTR:TT",...}
# path: 'books[*].someGuy'
"Bilbo"	"Gandalf"	"Sam"	"Aragorn"
# path: 'books[100]'
NULL
# path: 'books[0].isTrilogy'
false
# path: 'books[3].isTrilogy'
true
# path: 'books[number=1].title'
"LOTR:FOTR"
# path: 'books[title=LOTR:TT].double'
2
# path: 'books[double=2.0]'
{"title": "LOTR:TT",...}
# path: 'books[other=null].number'
0
# path: 'books[isTrilogy=true]'
{"title": "LOTR:FOTR",...}	{"title": "LOTR:TT",...}	{"title": "LOTR:ROTK",...}
# path: 'books[isTrilogy=false]'
{"title": "hobbit",...}
# path: 'extra.extra[0]'
3209423

